tests/utils/test_crypt_util.py

```diff
@@ -0,0 +1,75 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+import base64
+import binascii
+import errno as errno
+import glob
+import random
+import string
+import subprocess
+import tempfile
+import uuid
+
+import azurelinuxagent.common.conf as conf
+import azurelinuxagent.common.utils.shellutil as shellutil
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.utils.cryptutil import CryptUtil
+from azurelinuxagent.common.exception import CryptError
+from azurelinuxagent.common.version import PY_VERSION_MAJOR
+from tests.tools import *
+from subprocess import CalledProcessError
+
+class TestCryptoUtilOperations(AgentTestCase):
+    def test_decrypt_encrypted_text(self):
+        encrypted_string = load_data("wire/encrypted.enc")
+        prv_key = os.path.join(self.tmp_dir, "TransportPrivate.pem") 
+        with open(prv_key, 'w+') as c:
+            c.write(load_data("wire/sample.pem"))
+        secret = ']aPPEv}uNg1FPnl?'
+        crypto = CryptUtil(conf.get_openssl_cmd())
+        decrypted_string = crypto.decrypt_secret(encrypted_string, prv_key)
+        self.assertEquals(secret, decrypted_string, "decrypted string does not match expected")
+
+    def test_decrypt_encrypted_text_missing_private_key(self):
+        encrypted_string = load_data("wire/encrypted.enc")
+        prv_key = os.path.join(self.tmp_dir, "TransportPrivate.pem")
+        crypto = CryptUtil(conf.get_openssl_cmd())
+        self.assertRaises(CalledProcessError, crypto.decrypt_secret, encrypted_string, "abc" + prv_key)
+    
+    def test_decrypt_encrypted_text_wrong_private_key(self):
+        encrypted_string = load_data("wire/encrypted.enc")
+        prv_key = os.path.join(self.tmp_dir, "wrong.pem")
+        with open(prv_key, 'w+') as c:
+            c.write(load_data("wire/trans_prv"))
+        crypto = CryptUtil(conf.get_openssl_cmd())
+        self.assertRaises(CalledProcessError, crypto.decrypt_secret, encrypted_string, prv_key)
+
+    def test_decrypt_encrypted_text_text_not_encrypted(self):
+        encrypted_string = "abc@123"
+        prv_key = os.path.join(self.tmp_dir, "TransportPrivate.pem") 
+        with open(prv_key, 'w+') as c:
+            c.write(load_data("wire/sample.pem"))
+        encrypted_string = u"abc@123"        
+        crypto = CryptUtil(conf.get_openssl_cmd())
+        if PY_VERSION_MAJOR <= 2:
+            self.assertRaises(TypeError, crypto.decrypt_secret, encrypted_string, prv_key)
+        else:
+            self.assertRaises(binascii.Error, crypto.decrypt_secret, encrypted_string, prv_key)
```

Comment:
Why is the error type different for Python 3?

Reply:
I don't know, I can get python 2 .7 or 2.6 to run tests locally, I just have to let the build do it.  Functionally it doesn't matter though.

---
In reply to: [194030253](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194030253) [](ancestors = 194030253)

Reply:
handling internally now and returning a CryptError for both.

---
In reply to: [194103617](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194103617) [](ancestors = 194103617,194030253)